src/vivarium/framework/components/manager.py

```diff
@@ -26,6 +26,60 @@ class ComponentConfigError(VivariumError):
     pass
 
 
+class OrderedComponentSet:
+    """A container for Vivarium components. It preserves ordering, enforces
+    uniqueness by name, and provides a subset of set-like semantics."""
+
+    def __init__(self, *args):
+        self.names = set()
+        self.components = []
+        if args:
+            self.update(args)
+
+    def add(self, component: Any) -> None:
+        if component in self:
+            raise ComponentConfigError(f"Attempting to add a component with duplicate name: {component}")
+        self.names.add(component.name)
+        self.components.append(component)
+
+    def update(self, components: Union[List, Tuple]):
+        for c in components:
+            self.add(c)
+
+    def __contains__(self, component: Any) -> bool:
+        if not hasattr(component, "name"):
+            raise ComponentConfigError(f"Component {component} has no name attribute")
+        return component.name in self.names
+
+    def __iter__(self) -> Iterator:
+        return iter(self.components)
+
+    def __len__(self) -> int:
+        return len(self.components)
+
+    def __bool__(self) -> bool:
+        return bool(self.components)
+
+    def __eq__(self, other) -> bool:
+        try:
+            if len(self) != len(other):
+                return False
+            for a, b in zip(self, other):
+                if a.name != b.name:
+                    return False
+            return True
+        except TypeError:
+            return False
+
+    def pop(self) -> Any:
```

Comment:
I'm wondering if it might make more sense to use a queue than a list for the components since you're always appending to the end and removing from the beginning. Maybe we don't expect to have enough components for the additional time required to remove from the beginning of a list to matter but it seems to fit the use case a little better?

Reply:
I don't think performance will be much of an issue because these lists are gonna stay relatively small orders of magnitude wise. It seems common to use lists as stacks and queues, they cover the same operations. I think the alternative would be a dequedeque but I'm ambivalent towards that, except that it's an extra import

Reply:
I think lists are commonly used as stacks but less so for queues - the list documentation says they're inefficient for that. but yeah probably not a huge performance issue so your call.